# memore

`MeMoRe` stand for Methylation Motif Refiner. This is a tool designed to validate methylation motifs detected with SMRT and ONT sequencing.

## Authors' notes
We are actively developing `MeMoRe` to facilitate usage and broaden features. All feedback is more than welcome. You can reach us on twitter ([@iamfanggang](https://twitter.com/iamfanggang) and [@AlanTourancheau](https://twitter.com/AlanTourancheau)) or directly through the [GitHub issues system](https://github.com/fanglab/MeMoRe/issues).

## Content
+ [Usage](#Usage)
+ [Tool showcase](#Tool-showcase)
  + [Analysis of SMRT results](#Analysis-of-SMRT-results)
  + [Analysis of ONT results](#Analysis-of-ONT-results)
+ [Documentation](#Documentation)
+ [Citation](#Citation)

## Usage
To apply `MeMoRe` on your own dataset or use our testing dataset please follow: https://fanglab-tools.shinyapps.io/MeMoRe/. `MeMoRe` is designed to be user freindly and therefore is distributed as a [Shiny] application through [shinyapps.io] hosting solution. `MeMoRe` can be run from any system, without computational resources restriction.

## Tool showcase
To showcase the toolbox applications and facilitate an understanding of the methods, we build-in example datasets for SMRT and ONT analyses in [MeMoRe app](https://fanglab-tools.shinyapps.io/MeMoRe/).

### Analysis of SMRT results

![Output SMRT](/docs/figures/todo.smrt.png "E. coli xxx methylation motif results")

### Analysis of ONT results

![Output ONT](/docs/figures/todo.ont.png "E. coli xxx methylation motif results")

## Documentation
For a comprehensive description of `MeMoRe` including a detailed tutorial, please consult the [complete documentation][Full Documentation].

## Citation

A preprint is in preparation. Meanwhile, please cite the GitHub repository: https://github.com/fanglab/MeMoRe.

[Shiny]: https://shiny.rstudio.com/
[shinyapps.io]: https://www.shinyapps.io/
[Full Documentation]: https://MeMoRe.readthedocs.io/en/latest/overview.html
